### ISSUE TITLE
Defining truncation from -2

### ISSUE THREAD

**Issue by m-yac** [agda/cubical#219]
Once #155 is merged I plan to make a PR adding n-truncated and n-connected maps (plus a few easy facts) and it would be nice to have a definition of truncation which is defined from -2.

I experimented with this here: https://github.com/m-yac/cubical/commit/309f3ed021be2ccf993f461ad9254a64039cb950

In that commit I defined `ℕ₋₂` in an analagous way to `ℕ₋₁` (along with literals for both) and used the following definition of truncation from negative two:
```agda
∥_∥_ : ∀ {ℓ} (A : Type ℓ) (n : ℕ₋₂) → Type ℓ
∥ A ∥ neg2  = Lift Unit
∥ A ∥ suc n = FromMinusOne.∥ A ∥ (1+ n) -- the old HIT definition
```

Since one needs to pattern match on `n` when doing anything with this definition, I don't think it is ideal (e.g. the definition of `|_|` is already clunky). I'm curious if anyone knows of a better one.

**Comment by m-yac** [comment]
If we define a strange notion of 'pointed sphere' we can define truncation from -2 in a single HIT...

Not sure if the pattern-matching convenience is worth having to motivate this new definition, I haven't seen this notion of 'pointed sphere' anywhere else.
```agda
-- conversion from decidable types to pointed types which is the identity on an already inhabited type

open import Cubical.Relation.Nullary
open import Cubical.Data.Empty

Ptd : ∀ {ℓ} (A : Type ℓ) → Dec A → Σ[ A∙ ∈ Type ℓ ] A∙
Ptd A (no  _) = Lift Unit , lift tt
Ptd A (yes x) = A , x

ap∙ : ∀ {ℓ ℓ'} {A : Type ℓ} {B : Type ℓ'} (d : Dec A)
      → (fst (Ptd A d) → B) → (A → B)
ap∙ (no ¬a) f a = ⊥-elim (¬a a)
ap∙ (yes _) f a = f a

-- define the type of 'pointed spheres' (??) using the above
-- (they could also be defined directly, but the above helps them make some sense)

open import Cubical.HITs.Sn hiding (S) renaming (S₊ to S)
open import Cubical.HITs.Susp

S-Dec : ∀ (n : ℕ₋₁) → Dec (S n)
S-Dec neg1 = no ⊥-elim
S-Dec (suc n) = yes north

S∙ : ℕ₋₁ → Type₀
S∙ n = fst (Ptd (S n) (S-Dec n)) -- S∙ neg1 :≡ Lift Unit  and  S∙ (suc n) :≡ S (suc n)

apS∙ : ∀ {ℓ} {A : Type ℓ} {n} → (S∙ n → A) → (S n → A)
apS∙ = ap∙ (S-Dec _) -- if n > -1, apS∙ f x :≡ f x  and if n ≡ neg1, apS∙ f x ≡ elim-⊥

-- alternative definition of ∥_∥_ over ℕ₋₂ using Ptd (S n)

data  ∥_∥'_ {ℓ} (A : Type ℓ) (n : ℕ₋₂) : Type ℓ where
  ∣_∣' : A  → ∥ A ∥' n
  hub : (f : S (1+ n) → ∥ A ∥' n) → ∥ A ∥' n
  spoke : (f : S∙ (1+ n) → ∥ A ∥' n) (x : S∙ (1+ n)) → hub (apS∙ f) ≡ f x

isContr∥∥₋₂ : ∀ {ℓ} {A : Type ℓ} → isOfHLevel zero (∥ A ∥' -2)
isContr∥∥₋₂ = hub _ , λ a → spoke (λ _ → a) (lift tt)

-- isOfHLevel (suc n) (∥ A ∥' (-2+ suc n)) should follow exactly as in #155
```

**Comment by ecavallo** [comment]
Another option is to instead use a definition of truncation as a localization (#138). There's a [direct definition as a HIT](https://homotopytypetheory.org/2011/12/06/inductive-localization/) or [a slightly less direct version](https://arxiv.org/pdf/1706.07526.pdf); I don't know which is more usable. In either case, that should give the correct type for -2 as well.

The constructors for the current definition of truncation are a subset of those for the definition of localization; for n > -2, it turns out that the extra constructors are redundant, but they make a difference for n = -2.

**Comment by m-yac** [comment]
Thanks for these references! Although the construction in the paper you linked is really neat, I suspect the easiest definition to work with would be a modified version of the one in the blog post, replacing the half-adjoint equivalence constructors with bi-invertible equivalence constructors:
```agda
data Localize (F : ∀ α → S α → T α) (A : Type ℓ) : Type _ where
  ∣_∣ : A  → Localize F A
  -- (_◦ F α) : (T α → X) → (S α → X) is a bi-invertible equivalence for all α
  extr : ∀ α → (S α → Localize F A) → (T α → Localize F A)
  extr-inv : ∀ α (h : S α → Localize F A) (s : S α) → extr α h (F α s) ≡ h s
  extl : ∀ α → (S α → Localize F A) → (T α → Localize F A)
  extl-inv : ∀ α (g : T α → Localize F A) (t : T α) → extl α (g ◦ F α) t ≡ g t
```
As with `BiInvInt`, one can prove `extr ≡ extl` and so on.

This suggests that we should use the following HIT for truncation from -2, but I found it really challenging to integrate this into the existing code in `Truncation.Properties` without just basically rewriting everything. (Let me know if anyone gives this a try!)
```agda
data  ∥_∥_ {ℓ} (A : Type ℓ) (n : ℕ₋₂) : Type ℓ where
  ∣_∣ : A  → ∥ A ∥ n
  hub : (f : S (1+ n) → ∥ A ∥ n) → ∥ A ∥ n
  spoke : (f : S (1+ n) → ∥ A ∥ n) (s : S (1+ n)) → hub f ≡ f s
  -- necessary for the n = -2 case, motivated by the definition of a localization
  hub' : (f : S (1+ n) → ∥ A ∥ n) → ∥ A ∥ n
  spoke' : (x : ∥ A ∥' n) → hub' (const x) ≡ x
```
At https://github.com/m-yac/cubical/commit/d3e6b20326a247cd0cf9c06866b5f6a1c4aed563 I tried out using a simplified version of the 'pointed sphere' definition I gave in the second comment on this issue, and it needed very few changes in `Truncation.Properties` to get working.

So in the interest of preserving the intuitive structure and proofs of #155 I think we should go with something close to what's in https://github.com/m-yac/cubical/commit/d3e6b20326a247cd0cf9c06866b5f6a1c4aed563 as a definition of truncation from -2; I'm planning on adding localization/nullifications in a later PR so the S^(n+1)-nullification definition can be proven equivalent there. I'd be interested to hear anyone's thoughts!

**Comment by ecavallo** [comment]
I think it's a good idea to go with your lightweight solution. For the most part the truncation is only used via the induction principles in `Truncation.Properties`, so we should be able to swap out implementations fairly easily later on if we decide we want to.

The only reason I'd want to use the localization definition as the "primary" one is that we'll probably end up proving the definition we use is equivalent to the localization in any case, and it can get annoying to be transporting theorems across these equivalences all the time. (We already have to do it with `∥ A ∥ 1` vs `∥ A ∥₁` and so on, but at least in that case there's some practical benefit to using more specialized one, which I don't think is the case here.)

**Comment by m-yac** [comment]
After a discouraging amount of failure trying to define dependent elimination* for nullification using bi-invertible and half-adjoint equivalence HIT definitions, I was finally able to do it using a path-split equivalence definition! As a nice consequence, taking this as the definition of truncation involves even *less* changes to `Truncation.Properties` than my proposed definition above!
```agda
data Null {ℓ ℓ'} (S : Type ℓ) (A : Type ℓ') : Type (ℓ-max ℓ ℓ') where
  ∣_∣ : A → Null S A
  -- the image of every map (S → Null S A) is contractible in Null S A
  hub   : (f : S → Null S A) → Null S A
  spoke : (f : S → Null S A) (s : S) → hub f ≡ f s
  -- the image of every map (S → x ≡ y) for x y : A is contractible in x ≡ y
  ≡hub   : ∀ {x y} (p : S → x ≡ y) → x ≡ y
  ≡spoke : ∀ {x y} (p : S → x ≡ y) (s : S) → ≡hub p ≡ p s
```
I went ahead and made a PR: #230

*by this I mean something of the following type:
```agda
ind : ∀ {ℓ ℓ' ℓ''} {S : Type ℓ} {A : Type ℓ'} {B : Null S A → Type ℓ''}
      → (nB : (x : Null S A) → isNull S (B x))
      → ((a : A) → B ∣ a ∣) → (x : Null S A) → B x
```
